[116] New `@ClientHeaderParam` annotation and TCK tests

andymc12 opened the pull request:
Includes spec document changes describing how to use `@ClientHeaderParam`, the annotation in the API (including javadoc) and TCK tests that test proper usage and exception conditions.

Signed-off-by: Andy McCright <j.andrew.mccright@gmail.com>

andymc12 commented:
This is part 1 of the solution for issue #116.

kenfinnigan commented on api/src/main/java/org/eclipse/microprofile/rest/client/annotation/ClientHeaderParam.java:
```diff
@@ -0,0 +1,94 @@
+ * When this annotation is placed at the interface level of a REST client interface, the specified header will be sent on each request for all
+ * methods in the interface.
+ * When this annotation is placed on a method, the header will be sent only for that method. If the same HTTP header is specified in an annotation
+ * for both the type and the method, only the header value specified in the annotation on the method will be sent.
```
Does this need to cater to header params that are lists?

For instance, if the annotation is present on interface and method, if the type is a list then we add them both?

andymc12 replied:
I think that an annotation on the method should always override an annotation on the interface - this is consistent with other patterns (like Java EE's security annotations).

But, I think that you make a good point about headers that are lists.  We probably should make the value of the annotation a `String[]` rather than a `String`.

kenfinnigan replied:
Could we not just recognize it's a multi valued header and combine them?

andymc12 replied:
I think it would add complexity.  Potentially any header could have multiple values, so then we'd run into issues where it isn't clear whether a user is trying to override a header value in the method or whether they are trying to append another value.  We could add some sort of "appendable" attribute to the annotation to indicate that the value should be appended if multiple headers are already defined, but I think that makes it more complicated - and potentially unwieldy in a case like this:

```
@ClientHeaderParam(name="MyHeader", value="FromType")
public interface MyClient {

    @ClientHeaderParam(name="MyHeader", value="FromMethod")
    Response myMethod(@HeaderParam("MyHeader") passedInHeader);
}
...
myClient.myMethod("FromMethodArgument");
```

This could lead to sending:
`MyHeader: FromMethodArgument,FromType,FromMethod`
`MyHeader: FromType,FromMethod` (only using the ClientHeaderParams, ignoring the HeaderParam)
`MyHeader: FromMethodArgument` (only using the HeaderParam, ignoring the ClientHeaderParams)
or perhaps other combinations.

I was thinking that to make this as simple as possible, `@HeaderParam` takes priority over the `@ClientHeaderParam`, since that is passed-in by the user - part of the app.  Then the `@ClientHeaderParam` annotation on method would take priority over the annotation on the interface, since it is more specific.

So if a user wants multiple values for a particular header, then I think the annotation should support it with a `String[]` value attribute (and allow the default method to return a String array).  Alternatively, we could use multiple annotations at the same level - something like this:
```
@ClientHeaderParam(name="header1", value="value1")
@ClientHeaderParam(name="header1", value="value2")
@ClientHeaderParam(name="header1", value="value3")
Response myMethod();
```

but I prefer a String array attribute.

kenfinnigan replied:
Ok, I'd forgot about the override case.

So I'm good with making it a `String[]` and then following the original override means

kenfinnigan commented on api/src/main/java/org/eclipse/microprofile/rest/client/annotation/ClientHeaderParam.java:
```diff
@@ -0,0 +1,94 @@
+ * for both the type and the method, only the header value specified in the annotation on the method will be sent.
+ * <p>
+ * The value of the header to send can be specified explicitly by using the <code>value</code> attribute.
+ * The value can also be computed via a default method on the client interface that returns a String value.  This method must be specified in the
```
Are we also supporting {MyClass.myMethod} usage, or only default methods?

andymc12 replied:
My thought was only default methods.  That still gives users flexibility with extending other interfaces, etc.  Do you think we need to support {MyClass.myMethod} semantics?

kenfinnigan replied:
I think in the call we discussed utility methods too, so that you could have a common method to retrieve an authentication token, for instance, without having to replicate that code across all REST Client interfaces.

You could have the default method in multiple interfaces call that utility method, but that just seems like unnecessary boilerplate

andymc12 replied:
That makes sense.  So then the requirement for the compute method would be that it must either be a default method on the client interface, or a static method that is visible to the interface (i.e. not private, or package-protected in another package, etc.).

kenfinnigan replied:
Sounds good

kenfinnigan commented on spec/src/main/asciidoc/clientexamples.asciidoc:
```diff
@@ -80,6 +80,49 @@ All built in HTTP methods are supported by the client API.  Likewise, all base p
+While it is always possible to add a `@HeaderParam`-annotated method argument to specify headers, some times that does not make sense within the context of the application. For example, you may want to specify a username/password in the `Authorization` header to a secure remote service, but you may not want to have a `String authHeader` parameter in the client interface method.
+The `@ClientHeaderParam` annotation can allow users to specify HTTP headers that should be sent without altering the client interface method signature. The annotation contains three attributes: `name`, `value`. and `required`.  The name attribute is used to specify the header name. The value attribute is used to specify the value of the header. The value can be specified explicitly
+or it can reference a default method on the client interface that would compute the value of the header - in this latter case, the compute method name must be surrounded by curly braces. The compute method must be a default method on the interface, must return a `String` and must either contain no arguments or contain a single `String` argument - the implementation will use this
+`String` argument to pass the name of the header. The required attribute determines what should happen in the event that the compute method throws an exception. If the required attribute is set to true (default), then the client request will fail if the compute method throws an exception.  If it is set to false, then the client request will continue, but without sending the HTTP header.
```
Do we need some kind of big flashing warning in the spec doc about the fact that required=true will cause an exception if the method call fails?

andymc12 replied:
Sure, I can bold that text.

andymc12 commented:
@kenfinnigan @michalszynkiewicz - can you take another look at this PR?  I've updated it with the review comments, including:
    - Highlighting `required` attribute behavior
    - Making `value` attribute a String[]
    - Enabling compute method to be from a different class
    - Validation failure if same header specified twice on same target

Thanks!

michalszynkiewicz commented on tck/src/main/java/org/eclipse/microprofile/rest/client/tck/interfaces/InvalidComputeMethodSignature.java:
```diff
@@ -0,0 +1,52 @@
+        return "should not be invoked - unexpected Integer arg";
+    }
+
+    default String invalidMethod(ClientRequestContext context, ClientRequestContext someOtherArg) {
```
should it work with single `ClientRequestContext`?
I cannot find a test that would check it.

andymc12 replied:
I was originally thinking that we could pass in a `ClientRequestContext` in addition the header name, but ultimately decided against it since a lot of the info in CRC would be invalid (or at least, not yet determined this early in the flow).  

I should remove that method from the TCK interface.

Is there anything in particular that you thought might be useful from the CRC?  We could possibly pass in the `Method` object from the client interface - that might allow compute methods to read other annotations, etc to better determine what the header value should be...

michalszynkiewicz replied:
I didn't have anything specific in mind.

michalszynkiewicz commented on spec/src/main/asciidoc/clientexamples.asciidoc:
```diff
@@ -80,6 +80,61 @@ All built in HTTP methods are supported by the client API.  Likewise, all base p
+While it is always possible to add a `@HeaderParam`-annotated method argument to specify headers, some times that does not make sense within the context of the application. For example, you may want to specify a username/password in the `Authorization` header to a secure remote service, but you may not want to have a `String authHeader` parameter in the client interface method.
+The `@ClientHeaderParam` annotation can allow users to specify HTTP headers that should be sent without altering the client interface method signature. The annotation contains three attributes: `name`, `value`. and `required`.  The name attribute is used to specify the header name. The value attribute is used to specify the value(s) of the header. The value can be specified explicitly
+or it can reference a method that would compute the value of the header - in this latter case, the compute method name must be surrounded by curly braces. The compute method must be either a default method on the interface or a public static method that is accessible to the interface, must return a `String` or `String[]` and must either contain no arguments or contain a single `String`
+argument - the implementation will use this `String` argument to pass the name of the header. When specifying a compute method as the value attribute, only one method may be specified - if more than one string is specified as the value attribute, and one of the strings is a compute method (surrounded by curly braces), then the implementation will throw a RestClientDefinitionException.
```
It would be good to also have a TCK test checking if `RCDException` is thrown with the case of "more than one string is specified as the value attribute, and one of the strings is a compute method"

andymc12 replied:
Agreed.  I'll add this check.

andymc12 commented on api/src/main/java/org/eclipse/microprofile/rest/client/annotation/ClientHeaderParam.java:
```diff
@@ -0,0 +1,94 @@
+ * for both the type and the method, only the header value specified in the annotation on the method will be sent.
+ * <p>
+ * The value of the header to send can be specified explicitly by using the <code>value</code> attribute.
+ * The value can also be computed via a default method on the client interface that returns a String value.  This method must be specified in the
```
Good catch.  I'll update this to be more clear that:
1) The method can be either a default method on the rest client interface or a public static utility method.
and
2) That the compute method can return String or String[].

andymc12 commented:
@michalszynkiewicz I updated the Javadoc to be consistent with the spec text to indicate that the compute method can be either a default method on the interface _or_ a public static method on another class/interface.  And I added a new test for cases where multiple values were specified on the same annotation and at least one uses the curly braces.  Thanks again for the review!

andymc12 commented:
@kenfinnigan I assume you are ok with the latest changes (based on discussion in gitter from a couple weeks ago).  I'll plan to merge tomorrow unless you have any objections.  Thanks (and Happy New Year)!
